python/pyarrow/_parquet.pyx

```diff
@@ -680,6 +680,9 @@ cdef class FileMetaData(_Weakrefable):
         -------
         are_equal : bool
         """
+        err_msg = lambda: ("Expected `other` to be of type FileMetaData"
+                           " but found " + str(type(other)))
```

Comment:
Wrap it in lambda so that we construct this only when the assertion fails.

Reply:
Why not just `if not isinstance…` and a regular `raise`? (assertions can technically be disabled.) And the idiomatic exception would be TypeError.

Finally, I think Cython already enforces the type due to the annotation above, so it's just that we need to check for None.

Reply:
> Finally, I think Cython already enforces the type due to the annotation above, so it's just that we need to check for None.

Indeed, and we can actually also let cython check for the None by changing the signature to `def equals(self, FileMetaData other not None):`

Reply:
> def equals(self, FileMetaData other not None):

I didn't know that. Thanks! Will try that and update!

Reply:
Updated :)